controllers/placementapi_controller.go

```
@@ -207,6 +207,19 @@ func (r *PlacementAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
 func (r *PlacementAPIReconciler) reconcileDelete(ctx context.Context, instance *placementv1.PlacementAPI, helper *helper.Helper) (ctrl.Result, error) {
 	util.LogForObject(helper, "Reconciling Service delete", instance)
 
+	// remove db finalizer
+	serviceDb, err := database.GetServiceDbByName(ctx, helper, instance.Name, instance.Namespace)
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+	if err == nil {
+		controllerutil.RemoveFinalizer(serviceDb, helper.GetFinalizer())
```

So lib-common adds the finalizer. Could you make it symmetric that lib-common does the removal too? I.e. we use the Database object for all the creation mechanism so lib-common hides how the MariaDBDatabase CR is handled. By doing "database.GetServiceDbByName" and getting back a MariaDBDatabase instance this abstraction becomes leaky. Can we move the finalizer removal also to the Database object to keep the abstraction?

---

@gibizer @stuggi how about this approach : 
- on the lib-common's side I've added the DeleteFinalizer func to the database module: 
https://github.com/pinikomarov/lib-common/blob/add_database_finalizers/modules/database/funcs.go#L262-L263
- Then in the placementapi_controller I've put a - database.DeleteFinalizer :
https://github.com/pinikomarov/placement-operator/blob/add_service_db_finalizer_deletion/controllers/placementapi_controller.go#L211

@stuggi I know we talked about returning a *Database obj for the DeleteFinalizer,
 but as the *Database struct has the *mariadbv1.MariaDBDatabase object in it:
https://github.com/pinikomarov/lib-common/blob/add_database_finalizers/modules/database/types.go#L29
```
type Database struct {
  database         *mariadbv1.MariaDBDatabase
  databaseHostname string
  databaseName     string
```
I'm hesitant to return it so we don't "leak" any unnecessary MariaDBDatabase instance as @gibizer put it, not sure about it though ..?

---

Returning a Database object that has a **private** mariadbv1.MariaDBDatabase field is totally OK as the client will not see / has no access to that private field. 

I'm not 100% what you would do with the returned Database instance in placement. If you plan to use it to call db.DeleteFinalizer() -on it, then be careful that such instance is created with NewDatabase but that is not part of  the code path of reconcileDelete(). Probably you can call NewDatabase in the reconcileDelete() code path too. So that is probably OK.

---

@stuggi , is there a reason to return a *Database in  https://github.com/pinikomarov/lib-common/blob/add_database_finalizers/modules/database/types.go#L29
?
we can just return an error (current commit) to handle any reconcile events ..